### DISCUSSION TITLE
Allow Intercept getServerSideProps

### DISCUSSION THREAD

**Discussion by pA1nD** [vercel/next.js#17515] [Ideas]
![Untitled-2020-09-28-0828](https://user-images.githubusercontent.com/8961100/94773687-4a309480-038a-11eb-8afd-7480969225dd.png)

The picture above shows a illustratory example of a access_token and refresh_token based authentication.
Currently there seems to be an issue when the access_token has expired and the refresh route needs to be triggered.

Important to know is that (in order to make sense) the refresh_token is stored only in the client and should only be available at /api/refresh . So it is only sent when a refresh is required.

Traditionally libraries like axios provie a solution via intercept, where it is possible to catch an error, trigger the token refresh and execute the request again.

It seems to be impossible to intercept the answer of getServerSideProps, and directly cause a refresh on the client.

Am I wrong? Did anyone solve this?

**Comment by pA1nD** [comment]
Here is some very early / basic implementation of a refresh_flow with SSR: https://github.com/pA1nD/next-getSSP-example

**Reply by eric-burel** [reply]
Very instructive work! We've been scratching our head on this for a while, your architecture sounds very good.

Ping @clement-faure

**Reply by eric-burel** [reply]
More broadly, does this pattern belong to `getServerSideProps` in the first place? 
Take the example of a dashboard app in a company. Say you have built the skeleton of your page without data, using `getServerSideProps`. You may still want to add authentication to this page, because you don't want unauthenticated random user to see the structure of your dashboard. 
Introducing `getServerSideProps` for this leads to all sort of issues, like not being able to export your dashboard anymore or to use `getStaticProps`. You don't want server side props, you want to handle authentication, but current architecture involves React and props and rendering architecture in the process for no reason.

So in this case, what you really need could be a middleware to Next server. See https://github.com/vercel/next.js/discussions/10437#discussioncomment-90459, I try to clarify the best approaches depending on we consider a public or a private application.

**Reply by kadirosmanust** [reply]
Perfect, thanks for the schema. I am working on redirecting users. Now I can understand how can i handle that process. Thanks both.
